### ISSUE TITLE
Parsing Typescript?

### ISSUE THREAD

**Issue by calebboyd** [meriyah/meriyah#7]
I know this was attempted with cherow. Is typescript parsing a goal with this greenfield project?

**Comment by nchanged** [comment]
I would love to see that!

**Comment by KFlash** [comment]
I'm not sure what the future brings. As you may know I'm retired and Meriyah is just my little experiment project where I try to do everything in a different way that haven't been done before. And also in the most craziest way to make sure things are fast as possible and match the specs 100%. 

So as it is now. I'm focusing on the ECMA standards, but I'm open for anything in the future. 

Regarding TS it should be easy enough to implement this with Meriyah due to it's internal modular design, but I'm still trying to come up with a few solutions on how to extend this parser or open it up somehow for 3rd party aka plugins. That would make things even easier. Maybe you have a few ideas about that?

As of now I still have my TODO list to finish

1) Edge cases
2) Scoping
3) Error reporting

After that I may look into other things again, but for now expect this process to go slow because I'm not always around.

**Comment by KFlash** [comment]
Just to mention it. Meriyah will be or still are 100% spec compat **only if* you turn on all the options. 
I'm now implementing lexical scoping and you would need to set the `lexical` option to true to get it working.

It is also worth mentioning that all parsers have implemented support for additional ECMAScript features for Web Browsers - popularly known as AnnexB. 

Meriyah doesn't support AnnexB out of the box simply because the Test262 suite doesn't have it either It's an extension of the language.

In this case just turn on the `webcompat` option and you are good to go :)

**Comment by KFlash** [comment]
@calebboyd To answer this question. Meriyah as a JS parser will not parse TS, but there will be a meriyah-ts parser in the future that will parse TS. You can find the repo [here](https://github.com/meriyah/meriyah-ts)

No time frame on this for now.

**Comment by KFlash** [comment]
@calebboyd You can track my progress [here](https://github.com/buntis/buntis) for a TypeScript parser. Everyone seems to want a TS parser, and there was  no way possible to extend Meriyah  to accept TS syntax so I'm developing a new parser from scratch.

**Comment by KFlash** [comment]
@calebboyd Still in need for a Typescript parser? You can check out [Buntis](https://github.com/buntis/buntis/issues/10) . Most of the TS syntax is already implemented, but it's still a WIP.

**Comment by aralroca** [comment]
In my case to support TS and Meriyah, I use Bun's transpiler:

https://bun.sh/docs/api/transpiler

before passing it to Meriyah. It's still faster than many parsers, but surely if done all at once it would still be faster. Besides the Bun transpiler is not available neither Node nor Deno.

**Comment by aralroca** [comment]
> In my case to support TS and Meriyah, I use Bun's transpiler:
> 
> https://bun.sh/docs/api/transpiler
> 
> before passing it to Meriyah. It's still faster than many parsers, but surely if done all at once it would still be faster. Besides the Bun transpiler is not available neither Node nor Deno.

Ok, I realized that Bun.Transpiler is not a good option because it removes the import attributes now supported by meriyah. 

Surely it wouldn't be bad to retake this feature, especially now that Node.js also supports TypeScript.

**Comment by 3cp** [comment]
We emit ESTree AST, but ESTree doesn't cover TypeScript syntax. That's the first obstacle.
The second obstacle is that we don't have the time/resource to do that.

For your usage, have you tried oxc parser https://oxc.rs? The ast is very close to ESTree.

**Comment by aralroca** [comment]
I had in mind a configuration to omit the types and the Estree would remain the same as before, the only difference being that meriyah would work without having to transpile TS to JS with another tool.

**Comment by kenbern** [comment]
Parsing TypeScript can be simple if you only need to remove the types. You can achieve this by iterating through each node, without the need for a dedicated parser. However, if your goal is to preserve TypeScript syntax within an abstract syntax tree (AST), the process becomes more complex. You'll need to handle specific TypeScript constructs like enum and const enum to prevent data loss. The Meriyah library isn't designed for this purpose and would require additional tools, such as a generator similar to Babel. The ESTree standard also doesn't include support for TypeScript, and there's no default AST standard for it. Adding full TypeScript support to Meriyah would be a very intricate undertaking.

**Comment by kenbern** [comment]
@3cp I'm onto this typescript challenge in my own code. I'm extending the pretty printer to handle typescript. Maybe time soon to swap out Prettier with a better tool?
[https://github.com/kenbern/Ecast/tree/main/eprint](https://www.google.com/url?sa=E&q=https%3A%2F%2Fgithub.com%2Fkenbern%2FEcast%2Ftree%2Fmain%2Feprint)

**Comment by kenbern** [comment]
Challenge completed!! "During the development process, I encountered the realization that Prettier already possesses the capability to format JSON and other languages. In response, I implemented a robust plugin system to enhance the extensibility of my pretty-printer.  The TypeScript syntax is now handled as a plugin and will be released soon, along with a dedicated JSON plugin. 

The undertaking proved to be technically challenging, but I successfully overcame the obstacles. However, I recognize that there is always room for improvement, and I acknowledge the superior expertise of individuals like @3cp in this domain. My current implementation should be considered a preliminary effort, open to further refinement and optimization."
